src/lib.rs

```diff
@@ -0,0 +1,127 @@
+#[deny(clippy::all)]
+///! Graph API Traits
+///
+use rand::SeedableRng;
+
+#[allow(dead_code)]
+/// A graph layer name.
+struct Layer(&'static str);
+
+/// Abstract object in a graph, eg. node or edge.
+trait GraphObject {
+    /// Identifier of all graph objects.
+    type Id;
+
+    /// Local graph object data.
+    type Data;
+
+    /// Return the edge id.
+    fn id(&self) -> Self::Id;
+
+    /// Get local data.
+    fn data(&self) -> &Self::Data;
+
+    /// Get mutable local data.
+    fn data_mut(&self) -> &mut Self::Data;
+}
+
+/// A graph  node.
+trait Node: GraphObject {
+    /// The edge type connected to this node.
+    type Edge: GraphObject;
+
+    /// Get connected nodes.
+    fn neighbors(&self) -> Iterator<Item = &Self>;
+
+    /// Get incoming and outgoing edges.
+    fn edges(&self) -> Iterator<Item = &Self::Edge>;
+}
+
+/// A graph edge between two nodes.
+trait Edge: GraphObject {
+    /// Get the edge weight.
+    fn weight(&self) -> f64;
+
+    /// The object from which the edge starts.
+    fn from(&self) -> <Self as GraphObject>::Id;
+
+    /// The object to which the edge points.
+    fn to(&self) -> <Self as GraphObject>::Id;
+}
+
+/// The Graph API
+trait GraphAPI<G: Graph> {
+    /// Entropy used to seed the CPRNG.
+    type Entropy;
+
+    /// Add a graph layer.
+    fn add_layer(&mut self, layer: Layer);
+
+    /// Remove a graph layer. The layer must not contain any nodes.
+    fn remove_layer(&mut self, layer: &Layer);
+
+    /// Return an immutable graph of the given layer.
+    fn graph(&self, layer: &Layer) -> Option<&G>;
+
+    /// Return the mutable graph of the given layer.
+    fn graph_mut(&mut self, layer: &Layer) -> Option<&mut G>;
+
+    /// Given initial entropy, returns a random seed.
+    fn seed<R: SeedableRng>(&mut self, entropy: Self::Entropy) -> R::Seed;
+}
+
+/// A graph of nodes and edges.
+trait Graph {
+    /// A graph node.
+    type Node: Node<Edge = Self::Edge>;
+
+    /// A graph edge between nodes.
+    type Edge: Edge;
+
+    /// Add a node to the graph on the specified layer.
+    fn add_node<N: Node>(&mut self, id: N::Id, data: N::Data);
```

Comment:
I gave a try to this idea of adding `N` as a type to the function, but this doesn't work. The code compiles but if you try writing an implementation, it crumbles, because you cannot annotate anywhere what `N` is. Be my guest and try this on the `osrank` side 😸

Reply:
Will try!

Reply:
Tried this and it seems to work fine. The osrank code is very generic so you need some annotations in a few places, but otherwise I didn't encounter a problem yet.

Reply:
Wow, neat! Do you have a pointer to some code I could look at? I have been wanting to do this myself but wasn't able to 😅

Reply:
For example:

```diff
-                        let neighbors = network.neighbours(&current_node);
-                        match neighbors.choose_weighted(&mut rng, |item| {
+                        let neighbors = network.neighbors(&current_node);
+                        match neighbors.choose_weighted(&mut rng, |item: &G::Edge| {
                             network
-                                .lookup_edge_metadata(&item.id)
-                                .and_then(|m| Some(get_weight(m)))
+                                .get_edge::<G::Edge>(item.id())
+                                .map(|e| e.weight())
                                 .unwrap()
                         }) {
                             Ok(next_edge) => {
-                                walk.add_next(next_edge.target.clone());
-                                current_node = next_edge.target;
+                                walk.add_next(next_edge.to::<G::Node>().clone());
+                                current_node = next_edge.to::<G::Node>();
                             }

```

Reply:
This uses the `Edge` trait I have, hence the `id()`

Reply:
@cloudhead right, but what about the `implementation` for `add_node` or `to`? The roadblock I did hit was not calling it, but actually provide a valid implementation, as the types won't match there 😢

Reply:
I have to try this in a separate mock implementation, will come back to you.

Reply:
Thanks dude, keep me posted 😊

Reply:
Ok, I think the reason it doesn't work is obvious in hindsight: with `from<N: Node>()` - it is universally quantified, so `N` needs to be a parameter of `Edge` - but of course that's not that nice. A much better way would be to move `Id` to the Graph level and share the type for both, since they should probably be using the same id type anyway.

Reply:
Indeed, this is why I nudged you to do the same exercise -- my head couldn't compute how that could have worked :) 

> A much better way would be to move Id to the Graph level and share the type for both, since they should probably be using the same id type anyway.

We could, although for testing is convenient _not to_: for nodes I want to use strings in my tests, for edges `usize`. Have a look at what I am talking about here:

https://github.com/oscoin/osrank-rs/blob/master/src/importers/csv.rs#L445-L453

This is also why my trait embed the `Id` at the `GraphObject` so that `Node` and `Edge` can both override it.